### PR TITLE
Fix: Revise the behavior under multiple PoW requests

### PR DESCRIPTION
**Pull request by Fangop**
* The code has been formatted by `gofmt`
* This fix require 2 commits to fix 2 different mistakes.
* The test multiple PoW requests test is added.

## Multiple PoW requests
The original C-GO interface will lead to errors under multiple PoW requests.
The 2 reasons are below:
* Create and destroy dcurl for every PoW, which leads to error access of dcurl
* Uncertain length of return trytes
These are fixed by 2 commit provided.

## fix(C-GO): Remove redundat initializations
This commit fixed:
* Create and destroy dcurl for every PoW, which leads to error access of dcurl
Change initialization to package `init`.
Added an api for destroy dcurl

## fix(C-GO): Align return trytes with length 2673
This commit fixed:
* Uncertain length of return trytes
The return trytes of dcurl_entry have uncertain length(2673 or 2674).
In C applications, dcurl will use [`init_trytes`](https://github.com/JulianATA/dcurl/blob/c9b1ce1becabeae4ad0db0dfcf270b24a6e76fc9/tests/test-dcurl.c#L67) to make sure the length is 2673.
However, using `init_trytres` may leads to memory leak in GO.
Hence I use a length constraint for fixing this, as the implementation of [`init_trytres`](https://github.com/DLTcollab/dcurl/blob/4a8b4bc3cf3f00aa496a8b6c9a2e761423779bb5/src/trinary.c#L77) using memcpy.

## feat(C-GO): Add mutiple PoW request test
This commit provide:
* A GO test for mutiple PoW request for SyncDcurlTest
Using goroutine, this test can simulate mutiple PoW request.
As the application of C-GO interface, this test verifies the usage of SyncDcurlProofOfWork.
The SyncDcurlProofOfWork should return correct `nonce` under multiple requests.

### CODE REVIEW

File: cgo/pow/pow_test.go [hunk 1]
```diff
@@ -26,3 +27,24 @@ func TestDcurlProofOfWork(t *testing.T) {
 	assert.True(guards.IsTransactionHashWithMWM(hashedTrytes, uint(mwm)))
 
 }
+
+func TestSyncDcurlProofOfWork(t *testing.T) {
+	waitingGroup := new(sync.WaitGroup)
+	waitingGroup.Add(10)
+
+	assert := assert.New(t)
+	t.Log("Use Sync dcurl to calculate 10 request of PoWs with 8 threads enabled and MWM = 8")
```
Comment (marktwtn) [inline cgo/pow/pow_test.go]: MWM = 14

Comment (marktwtn) [reply cgo/pow/pow_test.go]: requests of PoW

File: cgo/pow/pow.go [hunk 2]
```diff
@@ -11,6 +11,10 @@ extern bool dcurl_init();
 extern bool dcurl_destroy();
 extern int8_t *dcurl_entry(int8_t *trytes, int mwm, int threads);
 
+bool dcurl_init_wrapper(){
+	return dcurl_init(NULL);
```
Comment (marktwtn) [inline cgo/pow/pow.go:13]: Does it really have to use `NULL`?

**Review by marktwtn — CHANGES_REQUESTED** [review]

**Comment by marktwtn** [comment]
Where are the corresponding initialization and destroy functions in ` cgo/pow/pow_test.go`?

**Comment by marktwtn** [comment]
Fix commit message:
`Here is a test which simulates mutiple PoW request.`
mutiple -> multiple

File: cgo/pow/pow.go (hunk 2 above)
Comment (Fangop) [reply cgo/pow/pow.go:13]: It can be used to pass a configuration file in dcurl.

Comment (marktwtn) [reply cgo/pow/pow.go:13]: Sorry I forgot that `dcurl_init()` do need a pointer as its parameter.
I think `extern bool dcurl_init();` might need a little change?

**Comment by Fangop** [comment]
> Fix commit message:
> `Here is a test which simulates mutiple PoW request.`
> mutiple -> multiple

Resolved

File: cgo/pow/pow.go (hunk 2 above)
Comment (Fangop) [reply cgo/pow/pow.go:13]: I resolve it to `dcurl_init(void* config)`.

Comment (Fangop) [reply cgo/pow/pow.go:13]: Not really sure how to use the config file.

**Comment by marktwtn** [comment]
Fix commit message:
Remove `redundat` initializations -> Remove `redundant` initializations 
Add multiple PoW `request`, simulates multiple PoW `request` -> PoW `requests`

File: cgo/pow/pow_test.go [hunk 3]
```diff
@@ -26,3 +27,24 @@ func TestDcurlProofOfWork(t *testing.T) {
 	assert.True(guards.IsTransactionHashWithMWM(hashedTrytes, uint(mwm)))
 
 }
+
+func TestSyncDcurlProofOfWork(t *testing.T) {
+	waitingGroup := new(sync.WaitGroup)
+	waitingGroup.Add(10)
+
+	assert := assert.New(t)
+	t.Log("Use Sync dcurl to calculate 10 requests of PoWs with 8 threads enabled and MWM = 14")
```
Comment (marktwtn) [inline cgo/pow/pow_test.go]: 10 requests of PoW

**Comment by Fangop** [comment]
> Fix commit message:
> Remove `redundat` initializations -> Remove `redundant` initializations
> Add multiple PoW `request`, simulates multiple PoW `request` -> PoW `requests`

Resolved?

**Review by marktwtn — CHANGES_REQUESTED** [review]
Rename the commit header
`feat(C-GO): Add multiple PoW requests test` -> `test(C-GO): Add multiple PoW requests test`

**Comment by Fangop** [comment]
> Rename the commit header
> `feat(C-GO): Add multiple PoW requests test` -> `test(C-GO): Add multiple PoW requests test`

Resolved.
Also, I added DcurlDestroy at the last of pow package test.

**Review by marktwtn — APPROVED** [review]

File: cgo/pow/pow.go [hunk 4]
```diff
@@ -71,8 +72,9 @@ func DcurlEntry(trytes Trytes, mwm int, parallelism ...int) (Trytes, error) {
 	}
 
 	returnTrytesC := C.dcurl_wrapper(C.CString(trytes), C.int(mwm), C.int(numThread))
-	returnTrytesGO := C.GoString(returnTrytesC)
-	nonce := returnTrytesGO[len(returnTrytesGO)-NonceTrinarySize/3-1 : len(returnTrytesGO)-1]
+	returnTrytesGO := C.GoString(returnTrytesC)[:2673]
```
Comment (jserv) [inline cgo/pow/pow.go]: Add comments for the following constants.

Comment (Fangop) [reply cgo/pow/pow.go]: I resolved the constant 2673 in line 75 to `TransactionTrytesSize` from [iota.go/consts](https://github.com/iotaledger/iota.go/blob/3dfaebe62d8b4037de292cd9c0bbd1a6027a9873/consts/consts.go#L133) for it is more concise than adding comment.

And, I added comment for the fallowing behavior of extracting nonce from trytes.

**Pull request merged**